src/Sylius/Bundle/CoreBundle/Resources/config/services.xml

```diff
@@ -64,15 +64,22 @@
         <service id="Sylius\Component\Core\Payment\InvoiceNumberGeneratorInterface" alias="sylius.invoice_number_generator" />
 
         <service id="sylius.image_uploader" class="Sylius\Component\Core\Uploader\ImageUploader">
+            <argument type="service" id="Sylius\Component\Core\Filesystem\Adapter\FilesystemAdapterInterface" />
+            <argument type="service" id="Sylius\Component\Core\Generator\ImagePathGeneratorInterface" />
+        </service>
+        <service id="Sylius\Component\Core\Uploader\ImageUploaderInterface" alias="sylius.image_uploader" />
+
+        <service id="Sylius\Component\Core\Filesystem\Adapter\FlysystemFilesystemAdapter">
+            <argument type="service" id="sylius.storage" />
+        </service>
+        <service id="Sylius\Component\Core\Filesystem\Adapter\GaufretteFilesystemAdapter">
             <argument type="service">
                 <service class="Gaufrette\Filesystem">
                     <argument>%sylius.uploader.filesystem%</argument>
                     <factory service="knp_gaufrette.filesystem_map" method="get" />
                 </service>
             </argument>
-            <argument type="service" id="Sylius\Component\Core\Generator\ImagePathGeneratorInterface" />
         </service>
```

Comment:
As far as I remember we wanted with @lchrusciel to keep the original shape of `sylius.image_uploader` service (Idk if that information has been missed when I hand over this task).

@lchrusciel take a look and give your feedback on what you think about it. If you decided the original shape must be preserved I'll call Kevin and tell him my initial idea how to solve it.

Reply:
The original shape is preserved but deprecated :P I just find it more suitable to use adapters instead of `AbstractImageUploader`. If you see the cons of this implementation, I'll open another PR 💃

Reply:
@coldic3 As far as I remember Łukasz asked me to do it this way but let's wait for his opinion 😄

Reply:
I was blind and stubborn with forcing different implementation. Current one is nice. @jakubtobiasz your first idea was good